From 92a00aeb0ffc8a60188ef731c9b47e1715f01432 Mon Sep 17 00:00:00 2001
From: Zahari Doychev <zahari.doychev@intel.com>
Date: Tue, 9 Jun 2020 15:17:23 +0200
Subject: [PATCH] dpm: ignore queue free for drop_q

DPM needs some enhancement to handle the drop_q and DP_SUBIF_SPECIFIC_Q. Add
some checks to avoid print errors in this cases for now.

Complete implementation will be provided in PONRTSYS-7430.
---
 drivers/net/datapath/dpm/gswip31/datapath_misc.c     | 3 ++-
 drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c | 2 +-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index 0423a9c0ca2c..1de2f24db9cc 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -1756,7 +1756,8 @@ static int deq_update_info(struct dp_subif_upd_info *info)
 		/* retrieve the qid corresponding to old deq port */
 		old_def_qid = old_deq_pinfo->first_qid;
 		old_deq_pinfo->ref_cnt--;
-		if (!free_queue_info(info->inst, info->dp_port, old_def_qid)) {
+		if (old_def_qid > 0 &&
+		    !free_queue_info(info->inst, info->dp_port, old_def_qid)) {
 			if (old_deq_pinfo->f_first_qid &&
 			    old_deq_pinfo->first_qid == old_def_qid) {
 				old_deq_pinfo->f_first_qid = 0;
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
index dd33ad9d2b1b..13cb4de9d3a6 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
@@ -3884,7 +3884,7 @@ int dp_queue_map_set_31(struct dp_queue_map_set *cfg, int flag)
 	}
 
 	if (!is_qid_valid(cfg->q_id) ||
-	    (is_q_node_free(priv, cfg->q_id) && cfg->q_id != priv->ppv4_drop_q))
+	    (cfg->q_id != priv->ppv4_drop_q && is_q_node_free(priv, cfg->q_id)))
 		return DP_FAILURE;
 
 	qmap_cfg.mpe1 = cfg->map.mpe1;
